### ISSUE TITLE
beego/adapter/config.NewConfig() crashed with an invalid xmlcontext

### ISSUE THREAD

**Issue by JX-Zhang98** [beego/beego#4698]
Please answer these questions before submitting your issue. Thanks!

1. What version of Go and beego are you using (`bee version`)?
go version go1.16.3 linux/amd64，
with beego commit id: 12af439a9c487f8ce3c561f4fe669f3ef2b3b0a4(latest at 2021/07/19)

2. What operating system and processor architecture are you using (`go env`)?
<details><summary><code>go env</code> </summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN="/home/lyf/workspace/gowork/bin"
GOCACHE="/home/lyf/.cache/go-build"
GOENV="/home/lyf/.config/go/env"
GOEXE=""
GOFLAGS="-mod=mod"
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/lyf/workspace/gowork/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/lyf/workspace/gowork"
GOPRIVATE=""
GOPROXY="https://goproxy.cn,direct"
GOROOT="/home/lyf/.local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/lyf/.local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.3"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/lyf/workspace/gowork/src/hwprojects/beego/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2259273923=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

3. What did you do?
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
I tried to create a config with xmlcontext, which is invalid. Here is my code:
<details><summary>beego/adapter/config/xml/Poc.go</summary><br><pre>
<code>
package xml

import (
	"os"
	"github.com/beego/beego/v2/adapter/config"
)

func PoC(data []byte) {
	var (
		xmlcontext = string(data)
	)
	cfgFileName := "testxml.conf"
	f, err := os.Create(cfgFileName)
	if err != nil {
		panic(err)
	}
	_, err = f.WriteString(xmlcontext)
	if err != nil {
		f.Close()
		panic(err)
	}
	f.Close()
	defer os.Remove(cfgFileName)
	xmlconf, err := config.NewConfig("xml", cfgFileName)
	if err != nil && xmlconf != nil {
		panic(err)
	} else if err != nil {
		return
	}
        return 
}
</code>
</details>

<details><summary>beego/adapter/config/xml/main/main.go</summary><br><pre>
<code>
package main
import (
	"github.com/beego/beego/v2/adapter/config/xml"
)

func main() {
        data = []byte{124, 99, 111, 110, 102, 105, 103, 62, 60, 97, 112, 112, 110, 97, 109, 101, 62, 97, 112, 105, 60, 47, 97, 112, 112, 110, 97, 109, 101, 62, 114, 116, 62, 60, 109, 121, 115, 113, 108, 112, 111, 114, 116, 62, 54, 48, 48, 60, 47, 109, 121, 115, 113, 108, 112, 111, 114, 116, 62, 60, 80, 73, 62, 57, 55, 54, 60, 47, 80, 73, 62, 60, 114, 117, 110, 109, 111, 100, 101, 62, 100, 101, 118, 60, 47, 114, 117, 110, 109, 111, 100, 101, 62, 60, 97, 117, 116, 111, 114, 101, 110, 100, 101, 114, 62, 108, 115, 101, 60, 47, 97, 117, 116, 111, 114, 101, 110, 100, 101, 114, 62, 60, 99, 111, 112, 121, 114, 101, 113, 117, 101, 115, 116, 98, 111, 100, 121, 62, 114, 117, 101, 60, 47, 99, 111, 112, 121, 114, 101, 113, 117, 101, 115, 116, 98, 111, 100, 121, 62, 60, 112, 97, 116, 104, 49, 62, 84, 72, 125, 60, 47, 112, 97, 116, 104, 49, 62, 60, 109, 121, 115, 101, 99, 116, 105, 111, 110, 62, 109, 101, 62, 60, 47, 109, 121, 115, 101, 99, 116, 105, 111, 110, 62, 60, 47, 99, 111, 110, 102, 105, 103, 62}
//("|config><appname>api</appname>rt><mysqlport>600</mysqlport><PI>976</PI><runmode>dev</runmode><autorender>lse</autorender><copyrequestbody>rue</copyrequestbody><path1>TH}</path1><mysection>me></mysection></config>")
        xml.PoC(data)
}
</code>
</details>

4. What did you expect to see?
Create the config according to xmlcontext successfully or fail with an err.

5. What did you see instead?
It crashed with error: "panic: interface conversion: interface {} is map[string]interface {}, not string"

**Comment by flycash** [comment]
Sorry, I can not reproduce this issue. Could you please update a demo? Your source code file. I copied the code, but cannot reproduce it.

**Comment by JX-Zhang98** [comment]
> Sorry, I can not reproduce this issue. Could you please update a demo? Your source code file. I copied the code, but cannot reproduce it.

This crash may due to an import to "github.com/beego/beego/v2/core/config/xml".

This PoC also works when "core" package is used.
Here is the new demo to trigger the crash:

```go

package main

import (
	"os"

	"github.com/beego/beego/v2/core/config"
	_ "github.com/beego/beego/v2/core/config/xml"
)

func main() {
	data := []byte{124, 99, 111, 110, 102, 105, 103, 62, 60, 97, 112, 112, 110, 97, 109, 101, 62, 97, 112, 105, 60, 47, 97, 112, 112, 110, 97, 109, 101, 62, 114, 116, 62, 60, 109, 121, 115, 113, 108, 112, 111, 114, 116, 62, 54, 48, 48, 60, 47, 109, 121, 115, 113, 108, 112, 111, 114, 116, 62, 60, 80, 73, 62, 57, 55, 54, 60, 47, 80, 73, 62, 60, 114, 117, 110, 109, 111, 100, 101, 62, 100, 101, 118, 60, 47, 114, 117, 110, 109, 111, 100, 101, 62, 60, 97, 117, 116, 111, 114, 101, 110, 100, 101, 114, 62, 108, 115, 101, 60, 47, 97, 117, 116, 111, 114, 101, 110, 100, 101, 114, 62, 60, 99, 111, 112, 121, 114, 101, 113, 117, 101, 115, 116, 98, 111, 100, 121, 62, 114, 117, 101, 60, 47, 99, 111, 112, 121, 114, 101, 113, 117, 101, 115, 116, 98, 111, 100, 121, 62, 60, 112, 97, 116, 104, 49, 62, 84, 72, 125, 60, 47, 112, 97, 116, 104, 49, 62, 60, 109, 121, 115, 101, 99, 116, 105, 111, 110, 62, 109, 101, 62, 60, 47, 109, 121, 115, 101, 99, 116, 105, 111, 110, 62, 60, 47, 99, 111, 110, 102, 105, 103, 62}
	//("|config>apirt>600976devlserueTH}me>")
	PoC(data)
}

func PoC(data []byte) {
	var (
		xmlcontext = string(data)
	)
	cfgFileName := "testxml.conf"
	f, err := os.Create(cfgFileName)
	if err != nil {
		panic(err)
	}
	_, err = f.WriteString(xmlcontext)
	if err != nil {
		f.Close()
		panic(err)
	}
	f.Close()
	defer os.Remove(cfgFileName)
	xmlconf, err := config.NewConfig("xml", cfgFileName)
	if err != nil && xmlconf != nil {
		panic(err)
	} else if err != nil {
		return
	}
	return
}

```

It crashed with :
```

2021/07/27 01:40:43.933 [W]  init global config instance failed. If you donot use this, just ignore it.  open conf/app.conf: no such file or directory

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 1 [running]:
github.com/beego/beego/v2/core/config/xml.(*Config).ParseData(0x963368, 0xc0001a2000, 0xd4, 0x200, 0x200, 0x0, 0x0, 0xc00018dda0)
        /home/jx/workspace/gowork/pkg/mod/github.com/beego/beego/v2@v2.0.1/core/config/xml/xml.go:74 +0x1b8
github.com/beego/beego/v2/core/config/xml.(*Config).Parse(0x963368, 0x73ee17, 0xc, 0x3, 0xc00014a3f8, 0x1, 0x0)
        /home/jx/workspace/gowork/pkg/mod/github.com/beego/beego/v2@v2.0.1/core/config/xml/xml.go:62 +0x96
github.com/beego/beego/v2/core/config.NewConfig(0x73c368, 0x3, 0x73ee17, 0xc, 0xd4, 0x0, 0x0, 0x178)
        /home/jx/workspace/gowork/pkg/mod/github.com/beego/beego/v2@v2.0.1/core/config/config.go:225 +0x82
main.PoC(0xc00018dea4, 0xd4, 0xd4)
        /home/jx/Desktop/crashes/xml/main.go:32 +0x1b4
main.main()
        /home/jx/Desktop/crashes/xml/main.go:13 +0x225
exit status 2

```

**Comment by JX-Zhang98** [comment]
Could you please apply a CVE ID for this vulnerability?
Thank you







------------------&nbsp;Original&nbsp;------------------
From:                                                                                                                        "beego/beego"                                                                                    ***@***.***&gt;;
Date:&nbsp;Tue, Aug 31, 2021 10:58 PM
***@***.***&gt;;
***@***.******@***.***&gt;;
Subject:&nbsp;Re: [beego/beego] beego/adapter/config.NewConfig() crashed with an invalid xmlcontext (#4698)





 
Closed #4698 via #4757.
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
